include/picongpu/plugins/saxs/Saxs.hpp

```diff
@@ -0,0 +1,815 @@
+                q_step = ( q_max - q_min ) /
+                         precisionCast< float_X >( numVectors );
+
+                // Rank 0 creates the output directory.
+                pmacc::Filesystem< simDim > & fs =
+                    Environment< simDim >::get( ).Filesystem( );
+                if ( isMaster )
+                {
+                    fs.createDirectory( "saxsOutput" );
+                    fs.setDirectoryPermissions( "saxsOutput" );
+                }
+
+                // Chose the solver for populating a TmpField with the electron
+                // density (either the species density or the bound electron
+                // density).
+                using ElectronDensitySolver =
+                    typename traits::DetermineElectronDensitySolver<
+                    T_ParticlesType >::type;
+                // Output unit:
+                const float_64 amplitudeUnit = static_cast< float_64 >(
+                    FieldTmp::getUnit< ElectronDensitySolver >( )[ 0 ] ) *
+                    CELL_WIDTH_SI * CELL_HEIGHT_SI * CELL_DEPTH_SI
+                    * ELECTRON_RADIUS_SI;
+
+                // Initialize an object responsible for output writing.
+                dataWriter = pmacc::memory::makeUnique< SaxsWriter< float_X > >(
+                    pluginPrefix + "Output",
+                    fileExtension,
+                    "saxsOutput",
+                    outputLayout,
+                    compressionMethod,
+                    precisionCast< uint64_t >( numVectors ),
+                    q_step,
+                    amplitudeUnit
+                );
+                // Set the output period.
+                outputPeriod = pluginSystem::toTimeSlice( outputPeriod_s );
+            }
+        }
+
+
+        void pluginUnload( ) override
+        {
+            if ( !notifyPeriod.empty( ) )
+            {
+                CUDA_CHECK( cudaGetLastError( ) );
+            }
+        }
+
+
+        //! Collect amplitude data from each CPU on the master node.
+        void collectIntensityOnMaster( )
+        {
+            auto const totalNumVectors = numVectors.productOfComponents( );
+
+            amplitude->deviceToHost( );
+            __getTransactionEvent( ).waitForFinished( );
+
+            reduce(
+                nvidia::functors::Add( ),
+                amplitudeMaster.data( ),
+                amplitude->getHostBuffer( ).getBasePointer( ),
+                bufferSize_i,
+                mpi::reduceMethods::Reduce( )
+            );
+
+        }
+
+
+        //! Calculates the offset to the the currently processed output chunk.
+        HINLINE uint32_t calcOffset( uint32_t const & step )
+        {
+            uint32_t totalRotations = accumulatedRotations + step;
+            uint32_t chunk = mpiRank + countRanks;
+            chunk = ( (chunk - ( totalRotations % countRanks ) ) %
+                       countRanks );
+            return chunk * bufferSize_i;
```

Comment:
This looks not obvious at all. Not saying it's wrong (honestly, I did not understand it from a first glance), but perhaps can be expressed more clearly, maybe with other variable names.

Reply:
Yeah I know it's very confusing :D  Basically what happens here:
* The chunks move with every "rotation" from left to the right (from smaller to a higher rank).
*  So for 6 GPUs we have:
     - in the begining  rank 0 has chunk 0, ..., rank 5 has chunk 5 
     - after first "rotation" 0 has 5, ... , 1 has 0, 2 has 1, ... 5 has 4 
     - and so on...
* So we have to subtract the number of rotation from the mpi rank and than usee modulo for the periodicity.
* To avoid negative numbers in the modulo operation (behaves a bit weirdly from what I remember) we add the number of ranks in the beginning and later subtract only `totalRotations % countRanks` since after every `countRanks` rotations we are anyway back at the initial setup.

A few examples:
 * for rank 0 after one rotation  : 0 + 6 = 6 -> 6 - (1 % 6) = 5 -> 5%6 = 5
 * after 7 rotations : 0 + 6 = 6 -> 6 - ( 7%6 ) = 5 -> 5%6 = 5
 * and for rank 5:  5 + 6 = 11 --> 11 -(7%6) = 10 -> 10 % 6 = 4  

I will try to put some comments for better understanding

Reply:
Thanks for explanations! Please do not resolve this one yet, as I would like to go through this code together with the explanation when reviewing the final version of the plugin.

Reply:
@sbastrakov what's the status on this comment ? Can I resolve it? :)

Reply:
Thanks for reminding. Just went through it again, looks good. Resolving.